runelite-client/src/main/java/net/runelite/client/plugins/config/ConfigPanel.java

```diff
@@ -205,11 +230,92 @@ void init(PluginConfigurationDescriptor pluginConfig)
 		rebuild();
 	}
 
+	private void toggleSection(ConfigDescriptor cd, JToggleButton button, JPanel contents, String sectionKey)
+	{
+		boolean newState = !contents.isVisible();
+		contents.setVisible(newState);
+		button.setIcon(newState ? SECTION_RETRACT_ICON : SECTION_EXPAND_ICON);
+		button.setRolloverIcon(newState ? SECTION_RETRACT_ICON_HOVER : SECTION_EXPAND_ICON_HOVER);
+		button.setToolTipText(newState ? "Retract" : "Expand");
+		pluginList.setConfigSectionExpandState(cd.getGroup().value(), sectionKey, newState);
+		SwingUtilities.invokeLater(contents::revalidate);
+	}
+
 	private void rebuild()
 	{
 		mainPanel.removeAll();
 
 		ConfigDescriptor cd = pluginConfig.getConfigDescriptor();
+
+		final Map<String, JPanel> sectionWidgets = new HashMap<>();
+		final Map<ConfigObject, JPanel> topLevelPanels = new TreeMap<>((a, b) ->
+			ComparisonChain.start()
+			.compare(a.position(), b.position())
+			.compare(a.name(), b.name())
+			.result());
+
+		for (ConfigSectionDescriptor csd : cd.getSections())
+		{
+			ConfigSection cs = csd.getSection();
+			final boolean isOpen = pluginList.getConfigSectionExpandState(cd.getGroup().value(), csd.getKey())
+				.orElse(!cs.closedByDefault());
+
+			final JPanel section = new JPanel();
+			section.setLayout(new BoxLayout(section, BoxLayout.Y_AXIS));
+			section.setMinimumSize(new Dimension(PANEL_WIDTH, 0));
+
+			final JPanel sectionHeader = new JPanel();
+			sectionHeader.setLayout(new BorderLayout());
+			sectionHeader.setMinimumSize(new Dimension(PANEL_WIDTH, 0));
+			// For whatever reason, the header extends out by a single pixel when closed. Adding a single pixel of
+			// border on the right only affects the width when closed, fixing the issue.
+			sectionHeader.setBorder(new CompoundBorder(
+				new MatteBorder(0, 0, 1, 0, ColorScheme.MEDIUM_GRAY_COLOR),
+				new EmptyBorder(0, 0, 3, 1)));
+			section.add(sectionHeader, BorderLayout.NORTH);
+
+			String name = cs.name();
+			final JLabel sectionName = new JLabel(name);
+			sectionName.setForeground(ColorScheme.BRAND_ORANGE);
+			sectionName.setFont(FontManager.getRunescapeBoldFont());
+			sectionName.setToolTipText("<html>" + name + ":<br>" + cs.description() + "</html>");
+			sectionHeader.add(sectionName, BorderLayout.WEST);
+
+			final JToggleButton sectionToggle = new JToggleButton(isOpen ? SECTION_RETRACT_ICON : SECTION_EXPAND_ICON);
```

Comment:
why are you using a toggle button, that doesn't have a toggled icon, that emulates toggling via some other method?

Reply:
It used to be an IconButton (I think that's what it was called) before that got removed, and when I looked at what replaced it, this was what I found

Reply:
You should either a) use the toggle functionality or b) not use a toggle button